packages/async-flow/src/replay-membrane.js

```diff
@@ -0,0 +1,389 @@
+      } else {
+        const entry = guestToHost(guestEntry);
+        log.pushEntry(entry);
+        const [_, ...args] = entry;
+        nestInterpreter(callIndex);
+        outcome = performCall(...args);
+      }
+    } catch (fatalError) {
+      throw panic(fatalError);
+    }
+
+    if (outcome.kind === 'return') {
+      return outcome.result;
+    } else {
+      outcome.kind === 'throw' ||
+        // @ts-expect-error TS correctly knows this case would be outside
+        // the type. But that's what we want to check.
+        Fail`unexpected outcome kind ${q(outcome.kind)}`;
+      throw outcome.problem;
+    }
+  };
+
+  // //////////////// Converters ///////////////////////////////////////////////
+
+  const makeGuestForHostRemotable = hRem => {
+    let gRem;
+    /** @param {PropertyKey} [optVerb] */
+    const makeGuestMethod = (optVerb = undefined) => {
+      const guestMethod = (...guestArgs) => {
+        const callIndex = log.getIndex();
+        return guestCallsHost(gRem, optVerb, guestArgs, callIndex);
+      };
+      if (optVerb) {
+        defineProperties(guestMethod, {
+          name: { value: String(optVerb) },
+          length: { value: Number(hRem[optVerb].length || 0) },
+        });
+      } else {
+        defineProperties(guestMethod, {
+          name: { value: String(hRem.name || 'anon') },
+          length: { value: Number(hRem.length || 0) },
+        });
+      }
+      return guestMethod;
+    };
+    const iface = String(getInterfaceOf(hRem) || 'remotable');
+    const guestIface = `${iface} guest wrapper`; // just for debugging clarity
+    if (typeof hRem === 'function') {
+      // NOTE: Assumes that a far function has no "static" methods. This
+      // is the current marshal design, but revisit this if we change our
+      // minds.
+      gRem = Remotable(guestIface, undefined, makeGuestMethod());
+      // NOTE: If we ever do support that, probably all we need
+      // to do is remove the following `throw Fail` line.
+      throw Fail`host far functions not yet passable`;
+    } else {
+      const methodNames = getMethodNames(hRem);
+      const guestMethods = methodNames.map(name => [
+        name,
+        makeGuestMethod(name),
+      ]);
+      // TODO in order to support E *well*,
+      // use HandledPromise to make gRem a remote presence for hRem
+      gRem = Remotable(guestIface, undefined, fromEntries(guestMethods));
+    }
+    return gRem;
+  };
+  harden(makeGuestForHostRemotable);
+
+  const makeGuestForHostVow = hVow => {
+    // TODO in order to support E *well*,
+    // use HandledPromise to make `promise` a handled promise for hVow
+    const { promise, resolve, reject } = makePromiseKit();
+    guestPromiseMap.set(promise, harden({ resolve, reject }));
+
+    watchWake(hVow);
+
+    void when(
+      hVow,
+      hostFulfillment => {
+        if (!log.isReplaying() && guestPromiseMap.get(promise) !== 'settled') {
```

Comment:
In what case would we have a guest already settled when we're not under replay? I can only imagine if somehow the host vow fulfilled twice, which would be erroneous behavior from the vow.

@erights you tried to remove these checks, which resulted in a behavioral difference and test failures. We should investigate why.

Reply:
The test is checking whether the promise was already resolved. When I turn it off
```suggestion
        if (!log.isReplaying() /* && guestPromiseMap.get(promise) !== 'settled' */) {
```
I get an extra log record
```js
    ['doFulfill', v1, 'x'],
```
at index 10 in the log that is redundant with the same
```js
    ['doFulfill', v1, 'x'],
```
at index 0. So the behavior with the check present is the correct behavior. 

The question is, why is it happening twice? I've managed to falsify every hypothesis I came up with. Gonna keep looking, but might need help.

Reply:
I got it. When the guest promise is created, it registers with the vow to get fulfilled once the vow is fulfilled. On replay, the vow is fulfilled, so when the guest promise for that vow is created, it registers. However, replay first fulfills that guest promise as instructed by interpreting the log's doFulfill. The but the `when` that registered the callback when the promise is created will still fire. The extra test is suppress reaction to the `when` callback.

So, regarding this issue, I'd say the code is correct as is. 

But walking through this makes me worried about the opposite case: If the vow is fulfilled, and the callback called, while we're replaying a log that has not recorded that this vow was settled. I think what we need to do is buffer those settling notifications and then react to them after replay is done. I think this is difficult to arrange because replay happens without really pausing. But each time the guest function does `await`s, it is an interleaving point. The callback might fire at that time, and the `!log.isReplaying()` above will cause that notification to get lost.

So, leaving this open until I either verify or falsify this opposite problem.

Reply:
Problem verified and fixed. I verified that the new test case at `test-replay-membrane-settlement.js` fails before the fix and passes after the fix.

The fix was easy. Testing it turned out to be surprising hard to figure out, but easy enough to write once figured out.

Done!